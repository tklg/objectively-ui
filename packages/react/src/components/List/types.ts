import { ForwardRefExoticComponent } from 'react'
import { PropsWithChildrenAndClassName } from 'src/types/PropsWithChildrenAndClassName'

export interface ListProps extends PropsWithChildrenAndClassName {
  ref?: React.RefObject<HTMLUListElement>;
  compact?: boolean;
  divided?: boolean;
  noPadding?: boolean;
}

export interface ListContextProps {
  compact: boolean;
  divided: boolean;
}

export type ListComponent<P = ListProps> = ForwardRefExoticComponent<P>
