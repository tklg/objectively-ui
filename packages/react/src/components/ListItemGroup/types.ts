import { ForwardRefExoticComponent } from 'react'
import { PropsWithChildrenAndClassName } from 'src/types/PropsWithChildrenAndClassName'

export type ListItemGroupProps = PropsWithChildrenAndClassName & {
  ref?: React.RefObject<HTMLLIElement>;
  subheading: string;
}

export type ListItemGroupComponent<P = ListItemGroupProps> = ForwardRefExoticComponent<P>
