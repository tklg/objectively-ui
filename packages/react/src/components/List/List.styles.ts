import { css } from '@emotion/react'
import { ColorTheme } from 'src/types/ColorTheme'
import { PROJECT_SHORTNAME } from 'src/utils/constants'

export const listStyles = (theme: ColorTheme) => css({
  listStyle: 'none',
  margin: 0,
  padding: `${theme.spacing.sm} 0`,

  [`&.${PROJECT_SHORTNAME}-List-noPadding`]: {
    padding: 0,
  },
})
