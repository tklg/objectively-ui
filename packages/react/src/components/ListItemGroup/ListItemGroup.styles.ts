import { css } from '@emotion/react'
import { ColorTheme } from 'src/types/ColorTheme'

export const listItemGroupStyles = (theme: ColorTheme) => css({
  padding: 0,
  margin: 0,
})

export const listItemGroupInnerListStyles = (theme: ColorTheme) => css({
  margin: 0,
  padding: 0,
  listStyle: 'none',
})

export const listItemGroupSubheadingStyles = (theme: ColorTheme) => css({
  padding: `${theme.spacing.md} calc(${theme.spacing.md} / 2) ${theme.spacing.xs} calc(${theme.spacing.md} / 2)`,
  margin: `0 calc(${theme.spacing.md} / 2)`,
  color: theme.colors.textSecondary,
  borderBottom: `${theme.lines.xs} solid ${theme.colors.divider}`,
  fontWeight: theme.typography.fontWeight.bold,
  fontSize: theme.typography.size.sm,

  ['ul li:first-of-type ul &']: {
    paddingTop: 0,
  },
})
