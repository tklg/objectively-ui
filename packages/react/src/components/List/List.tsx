import { createContext, forwardRef, useMemo } from 'react'
import { listStyles } from 'src/components/List/List.styles'
import { ListComponent, ListContextProps, ListProps } from 'src/components/List/types'
import { useTheme } from 'src/hooks'
import { buildClassName } from 'src/utils/buildClassName'

const ELEMENT_NAME = 'List'

export const ListContext = createContext<ListContextProps>({
  compact: false,
  divided: false,
})

export const List = forwardRef<HTMLUListElement, ListProps>(({
  children,
  className: _className,
  compact,
  divided,
  noPadding,
}, ref) => {
  const theme = useTheme()
  const className = buildClassName(ELEMENT_NAME, {
    compact,
    divided,
    noPadding,
  }, _className)

  const contextValue = useMemo(() => ({
    compact: compact ?? false,
    divided: divided ?? false,
  }), [compact])

  return (
    <ListContext.Provider value={contextValue}>
      <ul
        ref={ref}
        className={className}
        css={listStyles(theme)}
      >
        {children}
      </ul>
    </ListContext.Provider>
  )
}) as ListComponent

if (process.env.NODE_ENV !== 'production') {
  List.displayName = ELEMENT_NAME
}
