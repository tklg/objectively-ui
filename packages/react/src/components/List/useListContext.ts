import { useContext } from 'react'
import { ListContext } from 'src/components/List/List'

export const useListContext = () => useContext(ListContext)
