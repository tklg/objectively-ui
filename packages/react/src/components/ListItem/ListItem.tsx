import { css } from '@emotion/react'
import { forwardRef, useMemo } from 'react'
import { useListContext } from 'src/components/List/useListContext'
import { listItemButtonStyles, listItemStyles, listItemContentStyles, listItemTextStyles, listItemSubtextStyles } from 'src/components/ListItem/ListItem.styles'
import { ListItemComponent, ListItemProps } from 'src/components/ListItem/types'
import { useTheme } from 'src/hooks'
import { buildClassName } from 'src/utils/buildClassName'

const ELEMENT_NAME = 'ListItem'

export const ListItem = forwardRef<HTMLLIElement, ListItemProps>(({
  children,
  description,
  selected,
  button,
  className: _className,
  onClick,
}, ref) => {
  const { divided, compact } = useListContext()
  const theme = useTheme()
  const className = buildClassName(ELEMENT_NAME, {
    selected,
    button,
    divided,
    joined: !divided,
    compact,
    description: Boolean(description),
  }, _className)
  const listItemTextClassName = buildClassName(`${ELEMENT_NAME}Text`)
  const listItemTextDesciptionClassName = buildClassName(`${ELEMENT_NAME}Subtext`)

  const content = useMemo(() => (
    <div
      css={listItemContentStyles(theme)}
      data-selected={selected}
    >
      <div css={css({ width: '100%' })}>
        <span className={listItemTextClassName} css={listItemTextStyles(theme)}>{children}</span>
        {description && <span className={listItemTextDesciptionClassName} css={listItemSubtextStyles(theme)}>{description}</span>}
      </div>
    </div>
  ), [description, children, listItemTextClassName, listItemTextDesciptionClassName, theme])

  return (
    <li
      ref={ref}
      className={className}
      css={listItemStyles(theme)}
    >
      {button ? (
        <button
          css={listItemButtonStyles(theme)}
          onClick={onClick}
        >
          {content}
        </button>
      ) : (
        content
      )}
    </li>
  )
}) as ListItemComponent

if (process.env.NODE_ENV !== 'production') {
  ListItem.displayName = ELEMENT_NAME
}
