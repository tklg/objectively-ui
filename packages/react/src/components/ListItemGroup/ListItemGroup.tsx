import { forwardRef } from 'react'
import { useListContext } from 'src/components/List/useListContext'
import { listItemGroupInnerListStyles, listItemGroupStyles, listItemGroupSubheadingStyles } from 'src/components/ListItemGroup/ListItemGroup.styles'
import { ListItemGroupComponent, ListItemGroupProps } from 'src/components/ListItemGroup/types'
import { useTheme } from 'src/hooks'
import { buildClassName } from 'src/utils/buildClassName'

const ELEMENT_NAME = 'ListItemGroup'

export const ListItemGroup = forwardRef<HTMLLIElement, ListItemGroupProps>(({
  children,
  subheading,
  className: _className,
}, ref) => {
  const { compact } = useListContext()
  const theme = useTheme()
  const className = buildClassName(ELEMENT_NAME, {
    compact,
  }, _className)
  const innerListClassName = buildClassName(`${ELEMENT_NAME}InnerList`)
  const subheadingClassName = buildClassName(`${ELEMENT_NAME}Subheading`)

  return (
    <li
      ref={ref}
      className={className}
      css={listItemGroupStyles(theme)}
    >
      <ul className={innerListClassName} css={listItemGroupInnerListStyles(theme)}>
        <li className={subheadingClassName} css={listItemGroupSubheadingStyles(theme)}>{subheading}</li>
        {children}
      </ul>
    </li>
  )
}) as ListItemGroupComponent

if (process.env.NODE_ENV !== 'production') {
  ListItemGroup.displayName = ELEMENT_NAME
}
