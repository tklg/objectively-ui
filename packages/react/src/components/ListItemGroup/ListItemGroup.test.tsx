import { List } from 'src/components/List/List'
import { ListItem } from 'src/components/ListItem/ListItem'
import { ListItemGroup } from 'src/components/ListItemGroup/ListItemGroup'
import { render, screen } from 'src/__test__/render'

describe('ListItemGroup', () => {
  it('renders a list with a group and 1 item', async () => {
    render(
      <List data-testid='list'>
        <ListItemGroup subheading='Group 1'>
          <ListItem>Item 1</ListItem>
        </ListItemGroup>
      </List>,
    )

    expect(screen.getByTestId('list')).toMatchSnapshot()
  })
})
