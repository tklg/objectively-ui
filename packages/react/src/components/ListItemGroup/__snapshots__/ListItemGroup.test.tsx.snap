// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemGroup renders a list with a group and 1 item 1`] = `
<ul
  class="ObjUI-List emotion-0"
  data-testid="list"
>
  <li
    class="ObjUI-ListItemGroup emotion-1"
  >
    <ul
      class="ObjUI-ListItemGroupInnerList emotion-2"
    >
      <li
        class="ObjUI-ListItemGroupSubheading emotion-3"
      >
        Group 1
      </li>
      <li
        class="ObjUI-ListItem ObjUI-ListItem-joined emotion-4"
      >
        <div
          class="emotion-5"
        >
          <div
            class="emotion-6"
          >
            <span
              class="ObjUI-ListItemText emotion-7"
            >
              Item 1
            </span>
          </div>
        </div>
      </li>
    </ul>
  </li>
</ul>
`;
